### PR TITLE
[Mailer][Mime] Add PHPUnit constraints and assertions for the Mailer

### LINKED ISSUE #30850: [Mailer][RFC] KernelTestCase Assertion send messages
**Description**  
For functional tests it's sometimes necessary assert if some emails were "sent", for example a password reset or user registration email.

Currently there is no standard way to handle these assertions, for Swiftmailer in the past I used the profiler information but this required the profiler is actually enabled during testing. And it was less than optimal.

Now with the introduction of the Mailer component and improvements of the WebTestCase #30813 I propose the following assertions for email handling:

* `assertNoEmailsSent()`
* `assertEmailWasSent($messageAssertion)`

**Example** 
```php
$messageAssertion = new EmailAssertion(['john@example.com'])
    ->from('me@example.com')
    ->matchFrom('/me@example\.com/')

    ->subject('Your order was shipped')
    ->matchSubject('/Your order #\d+ shipped/s')

    ->text('Some text message')
    ->matchText('/Some text message/')

    //->hasAttachment('body', '/filename/', /* 'type' */)
    //->hasAttachmentFromPatch('source/file/path/image.jpg', '/filename/', /* 'type' */ )

    // priority, bcc, cc
;

$this->assertEmailWasSent($messageAssertion);
```

Alternative (automatically detect if arguments are an regex):

`assertEmailSent(['recipient list (including bcc, cc)'], 'subject', 'text', 'html');` 
Much lighter but also more arguments to remember.

-------

While the assertion is a bit verbose it's not possible to safely have multiple separate assertions as a priority assertion might match for one message but should match for another message instead. 

I don't think it makes sense add assertions for all possible information of a message though, the Mime Email class also allows attaching additional parts but this is more of an advanced use-case that would be fitted using a unit test for this specific usage.

The main focus of this proposal is assert if a message with some expected information was sent during the request cycle, message bus handling or Command handling.

### LINKED ISSUE #31747: [Mailer] NullTransport: store messages for verification in tests
**Description**  
When testing, the `NullTransport` in mailer is an attractive thing to use... because it completely disables delivery. But, if it *stored* the messages that were sent to it, then, in functional tests, we could easily fetch the transport from the container and verify its contents. This is what we do for the similar `InMemoruyTransport` in Messenger (https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Messenger/Transport/InMemoryTransport.php).

Another option would be to add info to the profiler and use that. I'm not sure which is better - it's always felt a bit awkward to need to enable the profiler whenever I needed to check if emails were sent, but that *is* also its job...

Relates to symfony/symfony#31592

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | refs #31592, closes #32409, closes #31947, closes #31747, closes #30850
| License       | MIT
| Doc PR        | n/a

This PR introduces PHPUnit constraints and assertions to ease testing emails in functional tests:

```php
<?php

namespace App\Tests;

use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;

class MailerAssertionsTraitsTest extends WebTestCase
{
    public function testSomething()
    {
        $client = static::createClient();
        $client->request('GET', '/');

        $this->assertEmailCount(2);
        $this->assertEmailIsQueued($this->getMailerEvent(0));

        $email = $this->getMailerMessage(0);
        $this->assertEmailHasHeader($email, 'To');
        $this->assertEmailHeaderSame($email, 'To', 'fabien@symfony.com');
        $this->assertEmailTextBodyContains($email, 'Bar');
        $this->assertEmailHtmlBodyContains($email, 'Foo');
        $this->assertEmailAttachementCount($email, 1);
    }
}
```

### CODE REVIEW

File: src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        $this->expectedValue = $expectedValue;
+        $this->transport = $transport;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toString(): string
+    {
+        return sprintf('%shas sent "%d" emails', $this->transport ? $this->transport.' ' : '', $this->expectedValue);
```
Comment (javiereguiluz) [inline src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php]: Is there a missing white space inside `%shas` ?

Comment (fabpot) [reply src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php]: No, the space is added when there is a transport name.

**Comment by fabpot** [comment]
image

**Comment by fabpot** [comment]
Now with some functional tests.

File: src/Symfony/Component/Mime/Test/Constraint/EmailAttachmentCount.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * @param RawMessage $message
+     *
+     * {@inheritdoc}
+     */
+    protected function matches($message): bool
+    {
+        if (RawMessage::class === \get_class($message) || Message::class === \get_class($message)) {
+            throw new \LogicException('Unable to test a message header on a RawMessage or Message instance.');
```
Comment (kevin-verschaeve) [inline src/Symfony/Component/Mime/Test/Constraint/EmailAttachmentCount.php]:
```suggestion
            throw new \LogicException('Unable to test a message attachment on a RawMessage or Message instance.');
```
?

File: src/Symfony/Component/Mime/Test/Constraint/EmailHtmlBodyContains.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * @param RawMessage $message
+     *
+     * {@inheritdoc}
+     */
+    protected function matches($message): bool
+    {
+        if (RawMessage::class === \get_class($message) || Message::class === \get_class($message)) {
+            throw new \LogicException('Unable to test a message header on a RawMessage or Message instance.');
```
Comment (kevin-verschaeve) [inline src/Symfony/Component/Mime/Test/Constraint/EmailHtmlBodyContains.php]:
```suggestion
            throw new \LogicException('Unable to test a message body on a RawMessage or Message instance.');
```
?

File: src/Symfony/Component/Mime/Test/Constraint/EmailTextBodyContains.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * @param RawMessage $message
+     *
+     * {@inheritdoc}
+     */
+    protected function matches($message): bool
+    {
+        if (RawMessage::class === \get_class($message) || Message::class === \get_class($message)) {
+            throw new \LogicException('Unable to test a message header on a RawMessage or Message instance.');
```
Comment (kevin-verschaeve) [inline src/Symfony/Component/Mime/Test/Constraint/EmailTextBodyContains.php]:
```suggestion
            throw new \LogicException('Unable to test a message body on a RawMessage or Message instance.');
```
?

File: src/Symfony/Bundle/FrameworkBundle/Test/MailerAssertionsTrait.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+     * @return MessageEvents[]
+     */
+    public static function getMailerEvents(string $transport = null): array
+    {
+        return self::getMessageMailerEvents()->getEvents($transport);
+    }
+
+    public static function getMailerEvent(int $index = 0, string $transport = null): ?MessageEvent
+    {
+        return self::getMailerEvents($transport)[$index];
```
Comment (kevin-verschaeve) [inline src/Symfony/Bundle/FrameworkBundle/Test/MailerAssertionsTrait.php]: is it missing `?? null` ?

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/Resources/config/routing.yml [hunk 6]
```diff
@@ -52,3 +52,7 @@ fragment_inlined:
 array_controller:
     path:     /array_controller
     defaults: { _controller: [ArrayController, someAction] }
+
+send_email:
+    path:     /send_email
+    defaults: { _controller: Symfony\Bundle\FrameworkBundle\Tests\Functional\Bundle\TestBundle\Controller\EmailController::indexAction, }
```
Comment (kevin-verschaeve) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/Resources/config/routing.yml]: is the trailing comma wanted ?

File: src/Symfony/Component/Mime/Test/Constraint/EmailAttachmentCount.php (hunk 2 above)
Comment (fabpot) [reply src/Symfony/Component/Mime/Test/Constraint/EmailAttachmentCount.php]: Good catch!

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/Resources/config/routing.yml (hunk 6 above)
Comment (fabpot) [reply src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/Resources/config/routing.yml]: all good catches, thank you, fixed now.

**Comment by javiereguiluz** [comment]
The proposed asserts are nice! Thanks.

I wonder if we need asserts for addresses. The mailer component allows to define email addresses in multiple ways (see https://symfony.com/doc/current/mailer.html#email-addresses). If I define some address as `Jane Smith <jane@example.com>` ... which of the following would work?

```php
$this->assertEmailHeaderSame($email, 'From', 'jane@example.com');
$this->assertEmailHeaderSame($email, 'From', 'Jane Smith');
$this->assertEmailHeaderSame($email, 'From', 'Jane Smith <jane@example.com>');
```

If others agree that this is confusing and also a common thing to assert, then we could add some assert for addresses which accept only a plain email address no matter how you defined the address:

```php
$this->assertEmailFromAddress($email, 'jane@example.com');
```

Related to this, what happens with multiple addresses?

```php
$this->assertEmailToAddressContains($email, 'john@example.com');
```

**Comment by fabpot** [comment]
@javiereguiluz Good idea. Implemented:

```php
        $this->assertEmailAddressContains($email, 'To', 'fabien@symfony.com');
        $this->assertEmailAddressContains($email, 'To', 'thomas@symfony.com');
        $this->assertEmailAddressContains($email, 'Reply-To', 'me@symfony.com');
```

Works with headers that can contain only 1 or several addresses (also with named addresses).

**Pull request merged**